### ISSUE TITLE
no-else-return : else if

### ISSUE THREAD

**Issue by bathos** [eslint/eslint#3015]
With no-else-return activated:

    if (x) {
      doOneThing();
    } else if (y) {
      return true;
    } else {
      doAnotherThing();
    }

The final else is considered a violation -- which makes technical sense given the implementation, but seems contrary to the intent of the rule (to eliminate superfluous else statements). The final else block is meaningful here despite the preceding "returning" if because it's also still an alternative of the first case, which did not return. Handling that distinction might be more trouble than it's worth, since grammatically, "else if" is not a concept, it's just "else statement" where the statement happens to be an "if", but I figured I'd ask just in case it's a viable refinement after all.

---

For anyone else running into this who'd like to keep the rule on -- this is how the no-else-return rule would prefer to see that logic expressed:

    if (x) {
      doOneThing();
    } else {
      if (y) return true;
      doAnotherThing();
    }

It's not that bad, just a bit unnatural when compared to the standard else-if pattern.

**Comment by nzakas** [comment]
Yeah, looks like a bug.

**Comment by wingedearth** [comment]
What about when you have just an if and else? Like:
```if (x) {
       return x;
     } else {
       throw new Error('no x');
     }```

I can't reverse the statement with !x because the linter reports no-negated-condition violation.

**Comment by platinumazure** [comment]
Simply write it like this:

```js
if (x) {
    return;
}

throw new Error();
```

If you have any other questions, please stop by our [chat](https://gutter.im/eslint/eslint).
